File: 201415/J154.pas
var
  h: array[1..100001] of int64;
  n, target, i: longint;
  money, missing, ffcycles, years: int64;
begin
  read(n, money);
  for i := 1 to n do
  begin
    read(h[i]);
    if (i >= 2) then
      if ((h[i - 1] = h[1]) and (money + h[1] < h[i])) then
      begin
        writeln('Impossible');
        halt
      end
  end;
  if (n <= 2) then
  begin
    writeln(0);
    halt
  end;
  money := money + h[1];
  h[n + 1] := 9999999999999999;
  target := 2;
  years := 0;
  while (true) do
  begin
    while (h[target] <= money) do
      inc(target);
    if (target = n + 1) then
    begin
      { check if can buy 1 year earlier}
      if ((money + h[1] >= h[n] * 2) and (years > 0)) then
        dec(years);
      writeln(years);
      halt
    end;
    missing := h[target] - money;
    { fast forward }
    ffcycles := (missing - 1) div (h[target - 1] - h[1]) div 2 + 1;
    if (ffcycles < 0) then
      ffcycles := 0;
    years := years + 4 * ffcycles;
    money := money + ffcycles * (h[target - 1] - h[1]) * 2
  end
end.

File: 201415/J154.cpp
#include <cstdio>
long long h[100002];
int main() {
  int n;
  long long money;
  scanf("%d %lld", &n, &money);
  for (int i = 1; i <= n; i++) {
    scanf("%lld", &h[i]);
    if (i >= 2 && h[i - 1] == h[1] && money + h[1] < h[i]) {
      printf("Impossible\n");
      return 0;
    }
  }
  if (n <= 2) {
    printf("0\n");
    return 0;
  }
  money += h[1];
  h[n + 1] = 9999999999999999LL;
  int target = 2;
  long long years = 0;
  while (true) {
    // find best home
    while (h[target] <= money) {
      target++;
    }
    if (target == n + 1) {
      // check if can buy 1 year earlier....
      if (money + h[1] >= h[n] * 2 && years > 0) {
        years--;
      }
      printf("%lld\n", years);
      return 0;
    }
    long long missing = h[target] - money;
    // fast foward
    long long ffcycles = (missing - 1) / (h[target - 1] - h[1]) / 2 + 1;
    if (ffcycles < 0) {
      ffcycles = 0;
    }
    years += 4 * ffcycles;
    money += ffcycles * (h[target - 1]-h[1]) * 2;
  }
}
